scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamJoiner.scala

```diff
@@ -0,0 +1,283 @@
+package com.twitter.scalding.beam_backend
+
+import com.twitter.scalding.serialization.Externalizer
+import com.twitter.scalding.typed.Joiner
+import com.twitter.scalding.typed
+import java.io.Serializable
+
+object BeamJoiner {
+  private def beamMapGroupJoin[K, R, R1](
+      fn: (K, Iterator[R]) => Iterator[R1]
+  ): (K, Iterable[R]) => Iterable[R1] = { (k: K, iter: Iterable[R]) =>
+    new Iterable[R1] {
```

Comment:
I think this is right, but the only risk is that it is unclear the default semantics of `Iterable`. Like, I think `map` will wind up materializing everything.

You could also experiment with using `Stream` instead, which will be lazy, but if someone holds the head it will possibly blow up the memory. I'm not sure the constraints you want.